sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java

```diff
@@ -200,17 +248,51 @@ public static Granularity convertSqlNodeToGranularity(SqlNode sqlNode) throws Pa
     throw makeInvalidPartitionByException(sqlNode);
   }
 
+  private static Granularity convertSqlLiteralCharToGranularity(SqlLiteral literal)
+  {
+    String value = literal.getValueAs(String.class);
+    try {
+      return Granularity.fromString(value);
+    }
+    catch (IllegalArgumentException e) {
+      try {
+        return new PeriodGranularity(new Period(value), null, null);
+      }
+      catch (Exception e2) {
+        throw makeInvalidPartitionByException(literal);
+      }
+    }
+  }
+
+  private static Granularity convertSqlIdentiferToGranularity(SqlIdentifier identifier)
```

Comment:
these 2 methods are almost identical; the difference is how they get to the `String` ; can you leave just one with a `String` arg - and do the X2String conversion before calling that method?